plonk/src/lagrange.rs

```diff
@@ -0,0 +1,228 @@
+//! Utilities for Lagrange interpolations, evaluations, coefficients for a
+//! polynomial
+
+use ark_ff::FftField;
+use ark_poly::{EvaluationDomain, Radix2EvaluationDomain};
+use ark_std::{ops::Range, vec, vec::Vec};
+
+// TODO: (alex) include these APIs upstream in arkworks directly
+
+/// A helper trait for computing Lagrange coefficients of an evaluation domain
+///
+/// from arkworks:
+/// Evaluate all Lagrange polynomials at tau to get the lagrange coefficients.
+/// Define the following as
+/// - H: The coset we are in, with generator g and offset h
+/// - n: The size of the coset H
+/// - Z_H: The vanishing polynomial for H. Z_H(x) = prod_{i in n} (x - hg^i) =
+///   x^n - h^n
+/// - v_i: A sequence of values, where v_0 = 1/(n * h^(n-1)), and v_{i + 1} = g
+///   * v_i
+///
+/// We then compute L_{i,H}(tau) as `L_{i,H}(tau) = Z_H(tau) * v_i / (tau - h *
+/// g^i)`
+#[allow(dead_code)]
+pub(crate) trait LagrangeCoeffs<F: FftField> {
+    /// Returns the first coefficient: `L_{0, Domain}(tau)`
+    fn first_lagrange_coeff(&self, tau: F) -> F;
+    /// Returns the last coefficient: `L_{n-1, Domain}(tau)`
+    fn last_lagrange_coeff(&self, tau: F) -> F;
+    /// Return a list of coefficients for `L_{range, Domain}(tau)`
+    fn lagrange_coeffs_for_range(&self, range: Range<usize>, tau: F) -> Vec<F>;
+}
+
+impl<F: FftField> LagrangeCoeffs<F> for Radix2EvaluationDomain<F> {
+    // L_0(tau) = Z_H(tau) * g^0 / (n * h^(n-1) *(tau - h * g^0))
+    // with g^0 = 1
+    // special care when tau in H, as both numerator and denominator is zero
+    fn first_lagrange_coeff(&self, tau: F) -> F {
+        let offset = self.coset_offset();
+        if tau == F::one() * offset {
+            return F::one();
+        }
+
+        let z_h_at_tau = self.evaluate_vanishing_polynomial(tau);
+        if z_h_at_tau.is_zero() {
+            // the case where tau is the first element in the coset
+            // already early-return
+            F::zero()
+        } else {
+            let denominator = self.size_as_field_element()
+                * offset.pow([self.size() as u64 - 1])
+                * (tau - offset);
+            z_h_at_tau * denominator.inverse().unwrap()
+        }
+    }
+
+    // L_n-1(tau) = Z_H(tau) * g^-1 / (n * h^(n-1) * (tau - h * g^-1))
+    // with g^n-1 = g^-1
+    fn last_lagrange_coeff(&self, tau: F) -> F {
+        let offset = self.coset_offset();
+        if tau == self.group_gen_inv() * offset {
```

Comment:
minor: we could early return if `tau` is either `offset` or `offset * g^{-1}`. Feels slightly weird that only one condition is considered in each case.

Reply:
i'll consider both first and last conditions in a new API `first_and_last` as you suggested

Reply:
I mean this could be in all 3 APIs.

Reply:
wait, i don't understand then. can you elaborate?

Reply:
oh no worry. my misunderstanding